### PR TITLE
fix(agent): point artifact reads at fq instead of absent jq

### PR DESCRIPTION
**Pull request by ljunb**
## Problem

A user asked 上下文呢; the agent called `recall_session_chunks`, whose output exceeded the inline budget and was spilled to a result artifact. Recovering it, the agent spent ~20 minutes re-reading that one file ~40 times:

```
python3 -c "... print(data['results'][0]['content'][:30000])"
python3 -c "... print(data['results'][0]['content'][:40000])"
...                                                  [:1800000]
```

Root cause is guidance, not logic. `prompt.go` offered `jq '.[0:20]' FILE` **ahead of** the equivalent `fq` form, but the board rootfs ships only `fq`, `yq`, `rg` (`scripts/rootfs_cli_tools.catalog`) — there is no `jq`. The model took the first example, hit `sh: jq: not found`, tried `import jq` (`ModuleNotFoundError`), then fell back to plain `json` with an escalating slice bound.

`fq` already evaluates jq expressions, so the right tool was present the whole time; only the guidance named the missing one. No new binary is needed.

## On-device verification

Board reached over the USB link (`192.168.42.1:8080`, agent daemon; SSH keys not provisioned):

```
fq -> /usr/bin/fq        0.17.0 (linux arm go1.26.0)
yq -> /usr/bin/yq        v4.53.3
rg -> /usr/bin/rg
jq -> MISSING            # jq ... -> sh: jq: not found, matching the log signature
```

`fq` runs jq syntax on ARMv7, on a `.data` artifact with no JSON extension, and the bound is real:

| command | result |
| --- | --- |
| `fq -r '.results[0].content[0:80]'` | 80 chars |
| `fq -r '.results[0].content[0:2000]' \| wc -c` | 2001 bytes (not the full 100KB) |
| `fq -r '.results \| length'` | `1` |
| `fq -r '.results[].chunk_id'` | `chunk_abc123` |

Probe file was created under `/tmp` and removed afterwards.

## Changes

- `prompt.go:158` — drop the `jq` example; state that `fq` evaluates jq expressions and `jq` is not installed; add an explicit bar on re-reading an artifact with a larger bound or falling back to `python3` for a growing slice.
- `tool_result_policy.go:358` — the inline recovery notice (the text sitting next to the artifact path when the model picks a command) said `grep/sed/dd/jq/fq`; now names `fq` only. This string is capped at 96 tokens including the path, so the detailed wording stays in the system prompt.
- Regression tests at both sites, asserting neither hands the model a bare `jq` command. Verified they fail when `jq` is reintroduced and pass when it is not.

## Testing

`go test ./internal/agent/ -count=1` passes (82s); `gofmt` and `go vet` clean.

## Not addressed

The loop guard still does not cover this shape. It blocked repeated `keyboard_tap` / `touch_gesture` / `quick_action` in the same log, but never fired across ~40 shell calls reading one file with an escalating bound. This PR stops the model from entering that loop; it does not make the loop detectable. Worth a follow-up.

Two side notes found while verifying, both left alone:
- Board `yq` v4.53.3 **does** support slicing (an earlier 4.44.3 does not) — `fq` is chosen because it speaks jq expressions natively, not because `yq` cannot slice.
- `prompt.go:129` authorizes `pip install --only-binary=:all:`, but PyPI `jq` has no armv7 wheel, so that path is a dead end on this hardware. It did not fire in this incident.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Updated artifact recovery guidance to use bounded `fq` and `yq` queries, byte ranges, and inline previews.
  - Clarified support for YAML inspection and improved handling of failed or uninformative queries.
  - Preserved useful previews and artifact locations when results are shortened due to context limits.
  - Improved detection of repeated artifact recovery attempts to stop potential loops.
  - Reset recovery tracking after new output, tool calls, or steering actions.

- **Tests**
  - Expanded coverage for detecting direct `jq` command usage across common command formats.
  - Verified bounded recovery behavior, complete short results, and loop prevention.

### CODE REVIEW

**Comment by aiden-lightning** [comment]
**Automated Claude Code review**

- PR: #657 — fix(agent): point artifact reads at fq instead of absent jq
- Head SHA: `c6ef5df389623e36edd954c64f60e34a9e141a56`
- GitHub updatedAt: `2026-09-09T02:17:41Z`
- Reviewed at: `2026-09-09T10:24:28+08:00`

## Summary

This PR fixes a production issue where the agent's system prompt suggested using `jq` for JSON processing, but `jq` is not installed on the target device (Aiden board). When the agent attempted to use `jq`, it would fail with "not found" and fall back to increasingly larger `python3` reads, creating infinite loops. The fix redirects all guidance to use `fq` (which is installed and evaluates jq expressions) and adds explicit anti-loop safeguards.

## Verified Issues

**None found.** The changes are well-designed and comprehensive.

## Observations

After reading the relevant code, I verified:

✅ **Consistent changes across all locations** - Both `prompt.go:158` and `tool_result_policy.go:358` are updated to remove `jq` references and clarify that only `fq` should be used

✅ **Comprehensive test coverage** - The new `TestRolePromptDoesNotSuggestUninstalledJqBinary` test checks for actual command patterns (`jq '`, `jq -r`, `| jq`, etc.) rather than just the word "jq", allowing the prompt to explain why jq shouldn't be used while preventing it from being suggested as a command

✅ **Anti-loop safeguards** - The expanded guidance explicitly warns: "never re-read the same artifact with a larger bound hoping for more content, and never fall back to python3 to print a growing slice" - directly addressing the failure mode described in commit 7b1f4001

✅ **Correct fq syntax** - The example `fq -r '.results[0].content[0:2000]' FILE` uses valid fq syntax for bounded JSON extraction

✅ **No orphaned references** - Verified via grep that the only jq references in agent code (excluding tests) were in the two files being modified

## Verdict

**Approve.** This is a targeted fix for a real production issue with appropriate test coverage. The changes prevent the agent from suggesting an unavailable command and provide clear guidance on the correct alternative, while adding safeguards against the infinite loop behavior.

File: src/agent/internal/agent/prompt_test.go [hunk 1]
```diff
@@ -90,6 +89,26 @@ func TestRolePromptConstrainsArtifactRecoveryReads(t *testing.T) {
...
 
+// The board rootfs ships fq/yq/rg but no jq, so guidance must never hand the
+// model a bare jq command: it fails with "not found" and the model then falls
+// back to python3 with an ever-growing slice bound, looping on one artifact.
+func TestRolePromptDoesNotSuggestUninstalledJqBinary(t *testing.T) {
+	profile := testPromptProfile(AgentConfig{})
+	for _, forbidden := range []string{"jq '", "jq -r", "jq \"", "| jq", "jq ."} {
+		if strings.Contains(profile.SystemPrompt, forbidden) {
+			t.Fatalf("system prompt suggests uninstalled jq binary via %q; use fq, which evaluates jq expressions:\n%s", forbidden, profile.SystemPrompt)
+		}
```
[the remark this answers is not in the document]
Comment (ljunb) [reply src/agent/internal/agent/prompt_test.go]: Valid — fixed in b33b8a14 (thread auto-resolved when the commented lines changed; replying for the record).

The literal list did let `jq -c FILE`, `jq --raw-output FILE`, and bare `jq FILE` through. Both sites now share a `findJqCommandForms` helper that matches on shape — an option, a quoted or bare filter, an uppercase operand, a pipe, or end of line — instead of enumerating forms, and `tool_result_policy_test.go` no longer uses the `/jq/` and `jq/fq` string checks.

One caveat worth flagging: my first draft keyed off leading shell separators (`^`, `|`, `;`, `&`) and was *weaker* than the literal list it replaced — it missed jq inside a comma-separated example list (`count=4096, jq -c FILE`) and inside a slash-separated tool list (`dd/jq/fq`), which are exactly the two shapes the original guidance used. The committed pattern covers both, and `TestJqCommandFormPatternMatchesInvocationsNotProse` pins the boundary in each direction so prose like "evaluates jq expressions" or "; jq is absent" stays allowed.

Verified by reintroducing each of the four jq forms into the source: every one fails both the prompt and the recovery-notice test, and all pass once reverted. Full package green (`go test ./internal/agent/ -count=1`, 79s).

**Comment by aiden-lightning** [comment]
**Automated Claude Code review**

- PR: #657 — fix(agent): point artifact reads at fq instead of absent jq
- Head SHA: `b33b8a149776e46ee04df7cab936fe23c1353c40`
- GitHub updatedAt: `2026-09-09T02:40:16Z`
- Reviewed at: `2026-09-09T10:50:38+08:00`

## PR Review: #657 - fix(agent): point artifact reads at fq instead of absent jq

### Summary

This PR removes references to the `jq` binary from agent prompts and tool result guidance, replacing them with `fq` (which evaluates jq expressions). The changes address a real issue where the agent was being instructed to use `jq`, which is not installed on the system, causing "command not found" errors and subsequent fallback loops.

### Verified Issues

**MEDIUM: Documentation still references jq without clarification**

The file `docs/02-architecture/paths-and-artifacts.md` contains jq command examples that readers might try to run on the device:

```
docs/02-architecture/paths-and-artifacts.md:119:jq . /userdata/agent/log/llm-http-$(date +%Y%m%d)*.log
docs/02-architecture/paths-and-artifacts.md:156:jq . /userdata/agent/log/llm-http-20260620090405123.log
docs/02-architecture/paths-and-artifacts.md:159:jq 'select(.kind | test("response|stream"))' llm-http-*.log
docs/02-architecture/paths-and-artifacts.md:162:jq -r 'select(.kind == "response") | .body | fromjson' llm-http-*.log
docs/02-architecture/paths-and-artifacts.md:165:jq -r '.kind' llm-http-*.log | sort | uniq -c
```

Since jq is not installed on the device rootfs (as this PR's commit message confirms), these documentation examples will fail when users try to run them. They should be updated to use `fq` instead, or include a note about the substitution.

**MEDIUM: Test infrastructure depends on jq availability**

The files `src/agent/internal/ota/manifest_test.go` and shell script `scripts/generate_ota_manifest.sh` explicitly require and use jq:

- `manifest_test.go:290` and `manifest_test.go:355`: `requireCommand(t, "jq")`
- `scripts/generate_ota_manifest.sh`: Uses jq extensively (15 occurrences)

The test uses `requireCommand()` which skips the test if jq is unavailable (line 408: `t.Skipf("%s not available: %v", name, err)`). This creates inconsistency: the agent runtime explicitly avoids jq because it's absent from the device, but the test/build infrastructure still depends on it.

While this doesn't break the agent at runtime, it means:
1. These tests will be skipped in environments without jq (like the device itself)
2. The build scripts require jq on the build machine, creating a dependency mismatch

This should be documented or the tooling should be migrated to fq for consistency.

### Observations

1. **Regex pattern is well-designed**: The `jqCommandFormPattern` correctly distinguishes between jq invocations (`jq '.[0:20]' FILE`) and prose mentions (`"jq is absent"`). The test coverage in `TestJqCommandFormPatternMatchesInvocationsNotProse` validates both positive and negative cases.

2. **Changes are consistent**: All three locations in the agent prompt code (prompt.go, tool_result_policy.go, and their tests) are updated cohesively.

3. **Test coverage is comprehensive**: The new tests ensure jq commands don't reappear in guidance and validate the detection regex thoroughly.

### Verdict

**APPROVE with recommendations**: The PR successfully fixes the agent guidance issue. The code changes are correct and well-tested. However, documentation and test infrastructure still reference jq in ways that could confuse users or create maintenance issues. Consider updating the documentation examples to use fq, or adding a note about the jq/fq distinction for developers vs. device environments.

**Comment by aiden-lightning** [comment]
**Automated Claude Code review**

- PR: #657 — fix(agent): point artifact reads at fq instead of absent jq
- Head SHA: `db0ecb1680fcbd5b0230c19d77b135194e1941b2`
- GitHub updatedAt: `2026-09-09T06:20:03Z`
- Reviewed at: `2026-09-09T14:28:59+08:00`

## PR Review: #657 - fix(agent): point artifact reads at fq instead of absent jq

### Summary

This PR addresses a critical agent loop issue where the model tries to use the `jq` binary (not installed) for artifact recovery, fails with "not found", then falls back to Python with progressively larger slice bounds, creating an infinite loop. The fix replaces `jq` references with `fq` (which IS installed and evaluates jq expressions) and adds termination policy to detect artifact recovery loops.

### Verified Issues

#### 🔴 HIGH: Inconsistent test expectation in prompt_test.go

**File**: `src/agent/internal/agent/prompt_test.go:83`

The current code (main branch) at line 83 shows `"jq"` is still in the expected list:
```go
for _, want := range []string{
    "grep -m 20",
    "sed -n",
    "dd if=FILE",
    "jq",  // ← Still here in current main
    "fq",
```

But the patch **removes** `"jq"` from this list in `TestRolePromptConstrainsArtifactRecoveryReads`. However, the actual prompt.go at line 158 still contains `jq '.[0:20]' FILE` as verified by my grep.

**Impact**: The test removal assumes prompt.go has been updated, but checking prompt.go line 158 shows it still contains the old jq command example. This means:
1. If the prompt.go change is applied but the test change isn't, the test will fail (won't find "jq")
2. The changes are interdependent and must be applied atomically

**Evidence**: My grep of prompt.go line 158 shows: `jq '.[0:20]' FILE, or fq '.[0:20]' FILE` still present in main branch.

**Verdict**: Not a bug in the PR itself, but confirms this is a multi-file atomic change that must be reviewed as a complete unit.

---

#### 🟡 MEDIUM: Potential context overflow with toolResultMinimumObservation

**File**: `src/agent/internal/agent/tool_result_policy.go:108`

The change modifies the inline threshold:
```go
contextLarge := availableTokens >= 0 && prepared.EstimatedTokens > max(toolResultMinimumObservation, availableTokens)
```

This means results under 96 tokens (`toolResultMinimumObservation`) will be returned inline even if `availableTokens` is much smaller (e.g., 10 tokens).

**Scenario**: If context budget has only 10 tokens available and a tool returns a 90-token error:
- Old behavior: Would artifact the result (because 90 > 10)
- New behavior: Returns inline (because 90 < 96)
- Result: Consumes 90 tokens against a 10-token budget = 80 token overflow

**Rationale per code**: The artifact recovery notice itself is ~96 tokens, so artifacting anything smaller would use MORE tokens than just returning it inline. This is a "least bad" option when context is nearly exhausted.

**Risk**: Could push context over hard limits in extreme budget exhaustion scenarios. No explicit overflow protection visible in the code path.

**Mitigation**: Test `TestToolResultPolicyKeepsShortResultsWhenContextIsFull` validates this behavior is intentional. The context manager may have buffers to handle small overruns.

**Confidence**: MEDIUM - The design rationale is sound, but edge cases with hard context limits aren't explicitly guarded.

---

### Observations

**Artifact Recovery Loop Detection** (termination_policy.go:272-290): The new logic tracks `artifactResultHash` and `artifactResultStreak` separately from the existing signature-based detection. This correctly handles alternating commands (python → cat → python) that all return the same error. The streak resets appropriately on steer, different tool calls, or new output. ✅

**Preview Prioritization** (tool_result_policy.go:343-367): Moving preview before metadata ensures useful error content survives when `combineMandatoryToolResultText` truncates to fit tight budgets. The test validates this with a 96-token budget. ✅

**jq Command Pattern Regex** (prompt_test.go:104): The pattern correctly distinguishes invocations (`jq -r FILE`, `dd/jq/fq`) from prose mentions (`"fq evaluates jq expressions"`, `"jq is absent"`). Verified against test cases. ✅

**Field Initialization**: `artifactResultHash` and `artifactResultStreak` are properly initialized in `ResetForSteer()` and reset in the `else` branch when non-artifact-recovery calls occur. ✅

---

### Verdict

**APPROVE with observation**

The PR correctly solves the jq loop problem with a multi-layered approach: eliminating bad guidance + detecting the loop pattern. The code changes are well-tested with 6 new test functions covering edge cases.

The `toolResultMinimumObservation` threshold change is intentional design—artifacting tiny results wastes more tokens than returning them inline. While this could theoretically overflow in extreme cases, the alternative (larger artifact recovery notices) would be worse. Consider documenting this trade-off in a code comment if not already present.

All changes work together as an atomic unit. The interdependencies between prompt.go, tool_result_policy.go, and their tests mean partial application would break functionality.

**Comment by aiden-lightning** [comment]
**Automated Claude Code review**

- PR: #657 — fix(agent): point artifact reads at fq instead of absent jq
- Head SHA: `1b89be9abfcd03e70ee6917c813da150a288b79b`
- GitHub updatedAt: `2026-09-09T07:21:49Z`
- Reviewed at: `2026-09-09T15:34:37+08:00`

Based on my review of PR #657, here's my analysis:

## Summary

This PR addresses a critical bug where the agent's artifact recovery guidance suggested using `jq` commands, but the board rootfs only ships with `fq`, `yq`, and `rg`. When the model tried to use `jq`, it would fail with "not found" and fall back to Python with ever-growing slice bounds, creating infinite loops. The PR fixes this by:

1. Replacing all `jq` command examples with `fq` 
2. Adding termination policy detection for artifact recovery loops
3. Improving artifact recovery guidance with more specific instructions
4. Ensuring tiny results stay inline to avoid unnecessary artifact recovery

## Verified Issues

### HIGH CONFIDENCE - Artifact recovery streak not reset in recordProgress()

**File**: src/agent/internal/agent/termination_policy.go:355-361

The `recordProgress()` method resets most loop detection state but **does not reset the artifact recovery streak**:

```go
func (p *TerminationPolicy) recordProgress() {
    p.stallScore = 0
    p.tier = TierNone
    p.lastNoticeTier = TierNone
    p.sameSigStreak = 1
    p.sameResultStreak = 1
    p.screenUnchangedAfterAction = 0
    // Missing: p.artifactResultHash = ""
    // Missing: p.artifactResultStreak = 0
}
```

**Impact**: When progress is detected (e.g., a successful web search), the artifact recovery streak persists. If the agent then returns to artifact recovery, it continues counting from the old streak rather than starting fresh. This could trigger a false positive loop detection.

**Evidence**: Compare with `ResetForSteer()` at lines 194-210, which explicitly resets `artifactResultHash` and `artifactResultStreak`. The test `TestTerminationPolicyResetsArtifactRecoveryStreak` verifies the reset behavior but only tests steer, other tools, and new output—it doesn't test the case where `recordProgress()` is called.

### MEDIUM CONFIDENCE - Potential interaction between artifact tracking and existing result tracking

**File**: src/agent/internal/agent/termination_policy.go:275-320

The artifact recovery tracking (lines 275-289) runs before the existing result hash tracking (lines 284-289). Both track `resultHash` but with different streak counters. When an artifact recovery call returns the same result:

- `artifactResultStreak` increments
- `sameResultStreak` also increments (line 285-289)

If `artifactResultStreak` reaches the limit first, it terminates with "artifact recovery calls repeated" (line 282). But if regular tracking reaches its limit first (line 322-324), it terminates with "same tool call repeated". The interaction seems correct, but the dual tracking means artifact recovery calls are counted twice, which could lead to earlier termination than expected for non-artifact recovery calls.

**Impact**: Artifact recovery calls are penalized more heavily than other repeated calls. This may be intentional defensive behavior, but it's worth confirming.

## Observations

1. **Test coverage is excellent**: The PR includes comprehensive tests for the regex pattern, artifact loop detection, reset behavior, and prioritization logic.

2. **The jq removal is thorough**: All `jq` command examples are replaced with `fq`, and a regex-based test prevents future reintroduction.

3. **Preview prioritization is sound**: Moving preview content before metadata in `boundedToolResultObservation` ensures useful output isn't displaced when space is tight.

4. **Minimum observation budget fix is correct**: Ensuring `toolResultMinimumObservation` applies to inline results prevents tiny results from being artifacted unnecessarily.

## Verdict

The PR solves the core problem effectively, but there's a verified defect in `recordProgress()` that should be fixed before merging. Add these two lines to the method:

```go
p.artifactResultHash = ""
p.artifactResultStreak = 0
```

The dual tracking interaction is likely acceptable but worth a second look to confirm it's intentional behavior.

**Review by HuangHan96 — APPROVED** [review]
